Add blog article about volume mode conversion going GA (#45262)
kubernetes/website: content/en/blog/_posts/2024-02-21-prevent-unauthorized-volume-mode-conversion-ga.md

---
layout: blog
title: "Kubernetes 1.30: Preventing unauthorised volume mode conversion moves to GA"

> date: 2024-03-22

Reviewer: The planned publication date for this blog is `2024-04-30`

Author: Fixed

slug: prevent-unauthorised-volume-mode-conversion-ga
---

**Author:** Raunak Pradip Shah (Mirantis)

With the release of Kubernetes 1.30, the feature to prevent the modification of the volume mode
of a [`PersistentVolumeClaim`](/docs/concepts/storage/persistent-volumes/) that was created from
an existing VolumeSnapshot in a Kubernetes cluster, has moved to GA!

### The problem

The [Volume Mode](/docs/concepts/storage/persistent-volumes/#volume-mode) of a `PersistentVolumeClaim` 

> whether the underlying volume on the storage device is formatted into a filesystem or presented as

Reviewer 2: The Volume Mode of a `PersistentVolumeClaim` refers to whether ...

Reviewer 3: Even better: no backticks

a raw block device to the Pod that uses it.

Users can leverage the `VolumeSnapshot` feature, which has been stable since Kubernetes v1.20,
to create a `PersistentVolumeClaim` (shortened as `PVC`) from an existing `VolumeSnapshot` in
the Kubernetes cluster. The `PVC` spec includes a `dataSource` field, which can point to an
existing `VolumeSnapshot` instance.
Visit [Create a PersistentVolumeClaim from a Volume Snapshot](/docs/concepts/storage/persistent-volumes/#create-persistent-volume-claim-from-volume-snapshot) 
for more details on how to create a `PVC` from an existing `VolumeSnapshot` in a Kubernetes cluster.

When leveraging the above capability, there is no logic that validates whether the mode of the
original volume, whose snapshot was taken, matches the mode of the newly created volume.

This presents a security gap that allows malicious users to potentially exploit an
as-yet-unknown vulnerability in the host operating system.

There is a valid use case to allow some users to perform such conversions. Typically, storage backup

> vendors convert the volume mode during the course of a backup operation for greater efficiency of

Reviewer 2: ... to retrieve changed blocks  for greater efficiency ...

Author: Done

operations. This prevents Kubernetes from blocking the operation completely and presents a challenge
in distinguishing trusted users from malicious ones.

> ### Preventing unauthorised users from converting the volume mode

Reviewer 3: ```suggestion
## Preventing unauthorised users from converting the volume mode
```

In this context, an authorised user is one who has access rights to perform `Update`
or `Patch` operations on `VolumeSnapshotContents`, which is a cluster-level resource.  

> It is upto the cluster administrator to provide these rights only to trusted users

Reviewer 4: ```suggestion
It is up to the cluster administrator to provide these rights only to trusted users
```

or applications, like backup vendors.

> Users apart from such unauthorised ones will never be allowed to modify the volume mode

Reviewer 4: Do you mean "from such _authorised_ ones" here, referring to the authorized backup vendors use case?

Author: Ooh good catch, fixed!

Reviewer 3: Although, US spelling, I think we prefer unauthorized

> In this context, an authorised user is one who has access rights to perform `Update`
> or `Patch` operations on `VolumeSnapshotContents`, which is a cluster-level resource.  
> It is up to the cluster administrator to provide these rights only to trusted users
> or applications, like backup vendors.
> Users apart from such authorised ones will never be allowed to modify the volume mode
> of a `PVC` when it is being created from a `VolumeSnapshot`.

Reviewer 3: Please update the article to match our style guide for API kinds (no backticks), and to write API verbs in lowercase bold:

```suggestion
In this context, an authorised user is one who has access rights to perform **update**
or **patch operations on VolumeSnapshotContents, which is a cluster-level resource.  
It is up to the cluster administrator to provide these rights only to trusted users
or applications, like backup vendors.
Users apart from such authorised ones will never be allowed to modify the volume mode
of a PVC when it is being created from a VolumeSnapshot.
```

Author: Thanks for the info, I've made that change here and elsewhere in the article.

To convert the volume mode, an authorised user must do the following:

1. Identify the `VolumeSnapshot` that is to be used as the data source for a newly
   created `PVC` in the given namespace.
2. Identify the `VolumeSnapshotContent` bound to the above `VolumeSnapshot`.

  ```shell
  kubectl describe volumesnapshot -n <namespace> <name>
  ```

> 3. Add the annotation [`snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"`](/docs/reference/labels-annotations-taints/#snapshot-storage-kubernetes-io-allowvolumemodechange)

Reviewer 2: following manifest fragment -> the following manifest fragment

Author: Done

>    to the above `VolumeSnapshotContent`. The `VolumeSnapshotContent` annotation must look like the following manifest fragment:

Reviewer 3: (nit)
```suggestion
   to the above VolumeSnapshotContent. The VolumeSnapshotContent annotations must include one similar to the the following manifest fragment:
```

  ```yaml
  kind: VolumeSnapshotContent
  metadata:
    annotations:
      - snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"
  ...
  ```

**Note**: For pre-provisioned `VolumeSnapshotContents`, you must take an extra
step of setting `spec.sourceVolumeMode` field to either `Filesystem` or `Block`,
depending on the mode of the volume from which this snapshot was taken.

An example is shown below:

   ```yaml
   apiVersion: snapshot.storage.k8s.io/v1
   kind: VolumeSnapshotContent
   metadata:
     annotations:

>      - snapshot.storage.kubernetes.io/allowVolumeModeChange: "true"

Reviewer 3: We should register this annotation before we release code for it, and we should try hard not to publish an announcement before the docs are all done.

What needs doing: update https://kubernetes.io/docs/reference/labels-annotations-taints/ on the _dev-1.30_ branch, to include a mention of this annotation.

Author: There was a mistake in the manifest fragment. The updated annotation already exists in the docs - https://kubernetes.io/docs/reference/labels-annotations-taints/#snapshot-storage-kubernetes-io-allow-volume-mode-change
I've made the change here

Reviewer 3: Ah, cool. That explains it - I did a string search for the annotation key as I saw it.

     name: <volume-snapshot-content-name>
   spec:
     deletionPolicy: Delete
     driver: hostpath.csi.k8s.io
     source:
       snapshotHandle: <snapshot-handle>
     sourceVolumeMode: Filesystem
     volumeSnapshotRef:
       name: <volume-snapshot-name>
       namespace: <namespace>
   ```

Repeat steps 1 to 3 for all VolumeSnapshotContents whose volume mode needs to be
converted during a backup or restore operation. This can be done either via software
with credentials of an authorised user or manually by the authorised user(s).

If the annotation shown above is present on a VolumeSnapshotContent object,
Kubernetes will not prevent the volume mode from being converted.
Users should keep this in mind before they attempt to add the annotation
to any VolumeSnapshotContent.

> 

Reviewer 2: We should add an Action Required Section to highlight what's in the "Urgent Upgrade Notes" sections of external-provisioner 4.0.0 and external-snapshotter v7.0.0:

https://github.com/kubernetes-csi/external-provisioner/releases/tag/v4.0.0
https://github.com/kubernetes-csi/external-snapshotter/releases/tag/v7.0.0

Reviewer 3: SGTM. Do you want that added to the main release announcement as well?

Reviewer: @sftim and @xing-yang I've added a note about this to the release blog. @RaunakShah are you able to add in a note about this?

### What's next

To determine which CSI external sidecar versions support this feature, please head
over to the [CSI docs page](https://kubernetes-csi.github.io/docs/).
For any queries or issues, join [Kubernetes on Slack](https://slack.k8s.io/) and

> create a thread in the #sig-storage channel. Alternately, create an issue in the

Reviewer 2: Can you add the #csi slack channel?

Author: Done

step of setting `spec.sourceVolumeMode` field to either `Filesystem` or `Block`,
depending on the mode of the volume from which this snapshot was taken.

An example is shown below:

   ```yaml
   apiVersion: snapshot.storage.k8s.io/v1
   kind: VolumeSnapshotContent
   metadata:
     annotations:
     - snapshot.storage.kubernetes.io/allow-volume-mode-change: "true"
     name: <volume-snapshot-content-name>
   spec:
     deletionPolicy: Delete
     driver: hostpath.csi.k8s.io
     source:
       snapshotHandle: <snapshot-handle>
     sourceVolumeMode: Filesystem
     volumeSnapshotRef:
       name: <volume-snapshot-name>
       namespace: <namespace>
   ```

Repeat steps 1 to 3 for all VolumeSnapshotContents whose volume mode needs to be
converted during a backup or restore operation. This can be done either via software
with credentials of an authorised user or manually by the authorised user(s).

If the annotation shown above is present on a VolumeSnapshotContent object,
Kubernetes will not prevent the volume mode from being converted.
Users should keep this in mind before they attempt to add the annotation
to any VolumeSnapshotContent.

> ## Action Required

Reviewer 3: (nit)
```suggestion
## Action required
```

The `prevent-volume-mode-conversion` feature flag is enabled by default in the 
external-provisioner `v4.0.0` and snapshot-controller `v7.0.0` CSI sidecar 
containers. Volume mode change will be rejected when creating a PVC from

> a VolumeSnapshot unless the steps described above have been performed.

Reviewer 2: One more comment: can you just change to "external-provisioner `v4.0.0` and external-snapshotter `v7.0.0`?  The flag is available in snapshot-controller but not really available in csi-snapshotter sidecar.
